conditional_camera/conditional_cam.go

```diff
@@ -0,0 +1,253 @@
+	images, meta, err := cc.cam.Images(ctx)
+	if err != nil {
+		return images, meta, err
+	}
+
+	extra, ok := camera.FromContext(ctx)
+	if !ok || extra[data.FromDMString] != true {
+		return images, meta, nil
+	}
+
+	for range images {
+		shouldSend, err := cc.shouldSend(ctx)
+		if err != nil {
+			return nil, meta, err
+		}
+
+		if shouldSend {
+			return images, meta, nil
+		}
+	}
+
+	cc.mu.Lock()
+	defer cc.mu.Unlock()
+
+	cc.addToBuffer_inlock(images, meta)
+
+	if len(cc.toSend) > 0 {
+		x := cc.toSend[0]
+		cc.toSend = cc.toSend[1:]
+		return x.imgs, x.meta, nil
+	}
+
+	return nil, meta, data.ErrNoCaptureToStore
+}
+
+func (cc *conditionalCamera) Stream(ctx context.Context, errHandlers ...gostream.ErrorHandler) (gostream.VideoStream, error) {
+	camStream, err := cc.cam.Stream(ctx, errHandlers...)
+	if err != nil {
+		return nil, err
+	}
+
+	return conditionalStream{camStream, cc}, nil
+}
+
+type conditionalStream struct {
+	cameraStream gostream.VideoStream
+	cc           *conditionalCamera
+}
+
+func (cs conditionalStream) Next(ctx context.Context) (image.Image, func(), error) {
+	extra, ok := camera.FromContext(ctx)
+	if !ok || extra[data.FromDMString] != true {
+		// If not data management collector, return underlying stream contents without filtering.
+		return cs.cameraStream.Next(ctx)
+	}
+
+	img, release, err := cs.cameraStream.Next(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	should, err := cs.cc.shouldSend(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if should {
+		return img, release, nil
+	}
+
+	cs.cc.mu.Lock()
+	defer cs.cc.mu.Unlock()
+
+	cs.cc.addToBuffer_inlock([]camera.NamedImage{{img, ""}}, resource.ResponseMetadata{CapturedAt: time.Now()})
+
+	return nil, nil, data.ErrNoCaptureToStore
+}
+
+func (cc *conditionalCamera) addToBuffer_inlock(imgs []camera.NamedImage, meta resource.ResponseMetadata) {
+	if cc.conf.WindowSeconds == 0 {
+		return
+	}
+
+	cc.cleanBuffer_inlock()
+	cc.buffer = append(cc.buffer, cachedData{imgs, meta})
+}
+
+func (cs conditionalStream) Close(ctx context.Context) error {
+	return cs.cameraStream.Close(ctx)
+}
+
+func (cc conditionalCamera) windowDuration() time.Duration {
+	return time.Second * time.Duration(cc.conf.WindowSeconds)
+}
+
+func (cc *conditionalCamera) cleanBuffer_inlock() {
+	sort.Slice(cc.buffer, func(i, j int) bool {
+		a := cc.buffer[i]
+		b := cc.buffer[j]
+		return a.meta.CapturedAt.Before(b.meta.CapturedAt)
+	})
+
+	early := time.Now().Add(-1 * cc.windowDuration())
+	for len(cc.buffer) > 0 {
+		if cc.buffer[0].meta.CapturedAt.After(early) {
+			return
+		}
+		cc.buffer = cc.buffer[1:]
+	}
+}
+
+func (cc *conditionalCamera) markShouldSend() {
+	cc.mu.Lock()
+	defer cc.mu.Unlock()
+
+	cc.captureTill = time.Now().Add(cc.windowDuration())
+	cc.cleanBuffer_inlock()
+
+	for _, x := range cc.buffer {
+		cc.toSend = append(cc.toSend, x)
+	}
+
+	cc.buffer = []cachedData{}
+}
```

Comment:
While you don't have to do this now, it's good to think about how this kind of logic will need to change when moving to the updated RDK version. There will still be buffers to capture, even if stream objects will not be around.